src/lib/storage/dictionary_column.hpp

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "base_attribute_vector.hpp"
+#include "fitted_attribute_vector.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+#include "value_column.hpp"
+
+namespace opossum {
+
+class BaseColumn;
+
+// Even though ValueIDs do not have to use the full width of ValueID (uint32_t), this will also work for smaller ValueID
+// types (uint8_t, uint16_t) since after a down-cast INVALID_VALUE_ID will look like their numeric_limit::max()
+constexpr ValueID INVALID_VALUE_ID{std::numeric_limits<ValueID::base_type>::max()};
+
+// Dictionary is a specific column type that stores all its values in a vector
+template <typename T>
+class DictionaryColumn : public BaseColumn {
+ public:
+  DictionaryColumn() = delete;
+
+  /**
+   * Creates a Dictionary column from a given value column.
+   */
+  explicit DictionaryColumn(const std::shared_ptr<BaseColumn>& base_column)
+      : _dictionary(nullptr), _attribute_vector(nullptr) {
+    auto value_column = std::dynamic_pointer_cast<ValueColumn<T>>(base_column);
+    DebugAssert(value_column != nullptr, "Type of ValueColumn does not match type of DictionaryColumn.");
+
+    const std::vector<T>& values = value_column->values();
+    std::set<T> sorter(values.cbegin(), values.cend());
+
+    _dictionary = std::make_shared<std::vector<T>>();
+    _dictionary->reserve(sorter.size());
+    std::copy(sorter.cbegin(), sorter.cend(), std::back_inserter(*_dictionary));
+
+    _attribute_vector = _create_attribute_vector(sorter.size());
+
+    for (const auto& value : values) {
+      const auto it = sorter.find(value);
+      _attribute_vector->set(_attribute_vector->size(), static_cast<ValueID>(std::distance(sorter.cbegin(), it)));
```

Comment:
you don't need a static_cast here: `ValueID(std::distance(sorter.cbegin(), it))`